File: 0001-Two-Sum.py
# 0001-Two-Sum.py

# Solution 1 time O(n) Space O(n)
def twoSum(self, nums: List[int], target: int) -> List[int]:
        ans = []
        l = len(nums)
        for i in range(l-1):
            for j in range(i+1, l):
                if nums[i]+nums[j] == target:
                    ans.append(i)
                    ans.append(j)
        return ans

# Solution 2 better solution
def twoSum(self, nums, target):
        d = {}
        for i, n in enumerate(nums):
            m = target - n          # m is the complement of n
            if m in d:
                return [d[m], i]
            else:
                d[n] = i

File: 0001_Two_Sum.java
// https://leetcode.com/problems/two-sum/discuss/3/Accepted-Java-O(n)-Solution
// use HashMap to record past nums and quickly find target in one loop
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] result = new int[2]; 
        Map<Integer, Integer> map = new HashMap<Integer, Integer>(); // () in the end
        for(int i =0; i < nums.length; i++){ //length no ()
            if(map.containsKey(target - nums[i])){ // containsKey
                result[1] = i;
                result[0] = map.get(target - nums[i]); // map get and put
                return result;
            }
            map.put(nums[i], i);
        }
        return result;
    }
}